src/core/vector/qgsvectorlayerundopassthroughcommand.cpp

```diff
@@ -189,16 +209,45 @@ void QgsVectorLayerUndoPassthroughCommandChangeGeometry::undo()
 {
   if ( rollBackToSavePoint() )
   {
-    emit mBuffer->geometryChanged( mFid, mOldGeom );
+    if ( mBuffer->mAddedFeatures.contains( mFid ) )
+    {
+      mBuffer->mAddedFeatures[ mFid ].setGeometry( mOldGeom );
+    }
+    else
+    {
+      QgsFeature tmp;
+      QgsFeatureRequest request;
+      request.setFilterFid( mFid );
+      request.setSubsetOfAttributes( {} );
+      std::unique_ptr<QgsVectorLayer> layerClone( layer()->clone() );
```

Comment:
This is **very** expensive to call. Why are we using a clone here instead of getting the features from the data provider directly?

Reply:
We cannot use the provider directly because it will "see" what is in the current transaction and not the status of the backend storage. What we need here is to know if the geometry has changed from what is in the storage outside of the current transaction.

I can't think of another way to access the status of the storage outside of the transaction than cloning the provider but suggestions are welcome.